### PR TITLE
Fast build check for test

### LINKED ISSUE #5108: Running unit tests aways does full build
When running unit tests, MD always rebuilds them. This is quite annoying when debugging. It should use the same fast "needs build" check that execution uses.

> VS bug [#636094](https://devdiv.visualstudio.com/DevDiv/_workitems/edit/636094)

### PR DESCRIPTION
**Pull request by mhutch**
Fixes #5108 Running unit tests aways does full build

### CODE REVIEW

File: main/src/addins/MonoDevelop.UnitTesting/Services/UnitTestService.cs [hunk 1]
```diff
@@ -126,32 +126,26 @@ internal static Task RunTest (UnitTest test, MonoDevelop.Projects.ExecutionConte
...
+					return;
 
-					var test_names = new HashSet<string> (tests.Select ((v) => v.FullName));
+				var test_names = new HashSet<string> (tests.Select ((v) => v.FullName));
 
-					await RefreshTests (cs.Token);
+				await RefreshTests (cs.Token);
 
-					tests = test_names.Select ((fullName) => SearchTest (fullName)).Where ((t) => t != null).ToList ();
+				tests = test_names.Select (SearchTest).Where ((t) => t != null).ToList ();
```
Comment (Therzok) [inline main/src/addins/MonoDevelop.UnitTesting/Services/UnitTestService.cs]: Please revert this change. Roslyn does crappy codegen for method groups.

**Review by Therzok — DISMISSED** [review]

**Review by slluis — DISMISSED** [review]

File: main/src/addins/MonoDevelop.UnitTesting/Services/UnitTestService.cs (hunk 1 above)
Comment (mhutch) [reply main/src/addins/MonoDevelop.UnitTesting/Services/UnitTestService.cs]: Huh. Is this filed with Roslyn?

File: main/src/addins/MonoDevelop.UnitTesting/Services/UnitTestService.cs [hunk 2]
```diff
@@ -126,32 +126,26 @@ internal static Task RunTest (UnitTest test, MonoDevelop.Projects.ExecutionConte
...
-						await IdeApp.ProjectOperations.CurrentRunOperation.Task;
-					}
 
-					foreach (var bt in build_targets) {
-						var res = await IdeApp.ProjectOperations.Build (bt, cs.Token).Task;
-						if (res.HasErrors)
-							return;
-					}
+				var res = await IdeApp.ProjectOperations.CheckAndBuildForExecute (
+					build_targets, ConfigurationSelector.Default, IdeApp.Preferences.BuildBeforeRunningTests,
```
[the remark this answers is not in the document]
Comment (mhutch) [reply main/src/addins/MonoDevelop.UnitTesting/Services/UnitTestService.cs]: yup

File: main/src/addins/MonoDevelop.UnitTesting/Services/UnitTestService.cs (hunk 1 above)
Comment (Therzok) [reply main/src/addins/MonoDevelop.UnitTesting/Services/UnitTestService.cs]: Yup, long standing too. https://github.com/dotnet/roslyn/issues/5835

Comment (Therzok) [reply main/src/addins/MonoDevelop.UnitTesting/Services/UnitTestService.cs]: https://github.com/dotnet/roslyn/issues/3563

File: main/src/core/MonoDevelop.Core/MonoDevelop.Projects/SolutionItem.cs [hunk 3]
```diff
@@ -635,35 +635,16 @@ public Task<BuildResult> Build (ProgressMonitor monitor, ConfigurationSelector s
...
-					referenced.RemoveAll (si => {
-						if (si is Project p)
-							return !p.FastCheckNeedsBuild (solutionConfiguration);
-						return false;//Don't filter things that don't have FastCheckNeedsBuild
-					});
-				var sortedReferenced = TopologicalSort (referenced, solutionConfiguration);
-
-				SolutionItemConfiguration iconf = GetConfiguration (solutionConfiguration);
-				string confName = iconf != null ? iconf.Id : solutionConfiguration.ToString ();
-				monitor.BeginTask (GettextCatalog.GetString ("Building: {0} ({1})", Name, confName), sortedReferenced.Count);
```
Comment (slluis) [inline main/src/core/MonoDevelop.Core/MonoDevelop.Projects/SolutionItem.cs]: This is changing the current behavior. Solution.BuildItems will log "Building Solution..." and the solution name, instead of "Building..." and the item name.

**Review by slluis — DISMISSED** [review]
Also, some tests are failing

File: main/src/core/MonoDevelop.Core/MonoDevelop.Projects/SolutionItem.cs (hunk 3 above)
Comment (mhutch) [reply main/src/core/MonoDevelop.Core/MonoDevelop.Projects/SolutionItem.cs]: Yup. Probably need Solution.BuildItems to log the name _if it's a single item_ and something generic otherwise?

File: main/tests/MonoDevelop.Core.Tests/MonoDevelop.Projects/ProjectBuildTests.cs [hunk 4]
```diff
@@ -350,8 +366,7 @@ string GetConfigFolderName (DotNetProject lib, string conf)
 		[Platform (Exclude = "Win")]
 		public async Task BuildWithCustomProps2 ()
 		{
-			string projFile = Util.GetSampleProject ("msbuild-tests", "project-with-custom-build-target2.csproj");
-			var p = (Project)await Services.ProjectService.ReadSolutionItem (Util.GetMonitor (), projFile);
+			(var sol, var p) = await LoadSampleSolutionItem<Project> ("msbuild-tests", "project-with-custom-build-target2.csproj");
```
Comment (slluis) [inline main/tests/MonoDevelop.Core.Tests/MonoDevelop.Projects/ProjectBuildTests.cs]: The solution is not disposed.

File: main/tests/MonoDevelop.Core.Tests/MonoDevelop.Projects/ProjectBuildTests.cs [hunk 5]
```diff
@@ -379,8 +394,7 @@ string GetConfigFolderName (DotNetProject lib, string conf)
 		[Platform (Exclude = "Win")]
 		public async Task BuildWithCustomProps3 ()
 		{
-			string projFile = Util.GetSampleProject ("msbuild-tests", "project-with-custom-build-target3.csproj");
-			var p = (Project)await Services.ProjectService.ReadSolutionItem (Util.GetMonitor (), projFile);
+			(var sol, var p) = await LoadSampleSolutionItem<Project> ("msbuild-tests", "project-with-custom-build-target3.csproj");
```
Comment (slluis) [inline main/tests/MonoDevelop.Core.Tests/MonoDevelop.Projects/ProjectBuildTests.cs]: Solution not disposed.

Comment (mhutch) [reply main/tests/MonoDevelop.Core.Tests/MonoDevelop.Projects/ProjectBuildTests.cs]: It is now :)

**Review by slluis — APPROVED** [review]

**Comment by slluis** [comment]
Thanks!
